sklearn/neighbors/tests/test_neighbors.py

```diff
@@ -163,6 +163,72 @@ def test_precomputed(random_state=42):
         assert_array_almost_equal(pred_X, pred_D)
 
 
+def test_precomputed_dense():
+    def make_train_test(X_train, X_test):
+        return (metrics.pairwise_distances(X_train),
+                metrics.pairwise_distances(X_test, X_train))
+
+    check_precomputed(make_train_test)
+
+
+@pytest.mark.parametrize('fmt', ['csr', 'lil'])
+def test_precomputed_sparse_implicit_diagonal(fmt):
+    def make_train_test(X_train, X_test):
+        nn = neighbors.NearestNeighbors(n_neighbors=3).fit(X_train)
+        return (nn.kneighbors_graph(None, mode='distance').asformat(fmt),
+                nn.kneighbors_graph(X_test, mode='distance').asformat(fmt))
+
+    check_precomputed(make_train_test, implicit_diagonal=True)
+
+
+def test_precomputed_sparse_explicit_diagonal():
```

Comment:
Do we really support an explicit diagonal in a precomputed sparse graph? I don't think we do, in which case I'm not sure this test is sufficient.

If we are keeping it, I think we can just parametrize implicit_diagonal